src/WarcraftLegacies.Source/Quests/Zandalar/QuestZulfarrak.cs

```diff
@@ -1,48 +1,68 @@
-using System.Collections.Generic;
+﻿using System.Collections.Generic;
 using MacroTools.Extensions;
 using MacroTools.FactionSystem;
 using MacroTools.QuestSystem;
 using MacroTools.QuestSystem.UtilityStructs;
-using MacroTools.Wrappers;
 using WarcraftLegacies.Source.Setup.Legends;
 using WCSharp.Shared.Data;
 using static War3Api.Common;
 
 
 namespace WarcraftLegacies.Source.Quests.Zandalar
 {
+  /// <summary>
+  /// Capture Zul'Farrak to unlock Gahz'rilla as a hero/>
+  /// </summary>
   public sealed class QuestZulfarrak : QuestData
   {
     private const int GAHZRILLA_ID = Constants.UNIT_H06Q_DEMIGOD_WARSONG;
-    private readonly List<unit> _rescueUnits = new();
-
+    private readonly List<unit> _rescueUnits;
+    private readonly List<unit> _killUnits;
+    /// <summary>
+    /// Initializes a new instance of the <see cref="QuestZulfarrak"/> class
+    /// </summary>
+    /// <param name="rescueRect"></param>
     public QuestZulfarrak(Rectangle rescueRect) : base("Fury of the Sands",
       "The Sandfury Trolls of Zul'farrak are openly hostile to visitors, but they share a common heritage with the Zandalari Trolls. An adequate display of force could bring them around.",
       "ReplaceableTextures\\CommandButtons\\BTNDarkTroll.blp")
     {
       ResearchId = Constants.UPGRADE_R02F_QUEST_COMPLETED_FURY_OF_THE_SANDS_WARSONG;
       AddObjective(new ObjectiveControlLegend(LegendNeutral.Zulfarrak, false));
-      foreach (var unit in new GroupWrapper().EnumUnitsInRect(rescueRect).EmptyToList())
-        if (GetOwningPlayer(unit) == Player(PLAYER_NEUTRAL_PASSIVE))
-        {
-          SetUnitInvulnerable(unit, true);
-          _rescueUnits.Add(unit);
-        }
+      AddObjective(new ObjectiveLegendReachRect(LegendTroll.LEGEND_PRIEST, rescueRect, "Zul'Farrak"));
+      _rescueUnits = rescueRect.PrepareUnitsForRescue(RescuePreparationMode.HideNonStructures);
+      _killUnits = rescueRect.PrepareUnitsForRescue(RescuePreparationMode.HideNonStructures, (unit => GetUnitTypeId(unit) != GetUnitTypeId(LegendNeutral.Zulfarrak.Unit)), Player(PLAYER_NEUTRAL_AGGRESSIVE));
```

Comment:
I think we want `RescuePreparationMode.None` for this line, as the trolls should start hostile and attackable.

Reply:
Yep true, missed that. Will adjust